src/data/matrix/companion.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2022 Pierre-Alexandre Bazin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pierre-Alexandre Bazin
+-/
+import ring_theory.adjoin_root
+import linear_algebra.matrix.to_lin
+
+/-!
+# Companion matrices
+
+## Main definitions
+
+* `matrix.companion` : the companion matrix of a polynomial.
+
+## Main statements
+
+* `matrix.companion_eq_to_matrix_lmul_root` : The companion matrix of `P` is the matrix of the
+  multiplication by `root P` in the canonical basis `1, root P, (root P)^2, ...` of `adjoin_root P`.
+
+## Tags
+
+Companion matrix, Frobenius matrix
+-/
+
+open_locale polynomial
+
+namespace matrix
+
+/--The companion matrix of a polynomial.-/
+def companion {R : Type*} [ring R] (P : R[X]) : matrix (fin P.nat_degree) (fin P.nat_degree) R :=
+if h : P.nat_degree = 0 then 0 else
+  λ i j, if j = fin.last' h then -P.coeff i else if (j : ℕ) + 1 = i then 1 else 0
```

Comment:
You can avoid the need for `last` with
```suggestion
match P.nat_degree with
| 0 := 0
| nat.succ n := λ i j, if j = fin.last _ then -P.coeff i else if (j : ℕ) + 1 = i then 1 else 0
end
```
or perhaps more conveniently
```suggestion
nat.cases_on P.nat_degree 0 $
  λ _ i j, if j = fin.last _ then -P.coeff i else if (j : ℕ) + 1 = i then 1 else 0
```

Reply:
It indeed makes things harder, because then to prove the lemma I want to do `cases P.nat_degree n` but then I don't have anything in the context that tells me `P.nat_degree = 0` or `P.nat_degree = n.succ`

Reply:
`cases h : P.nat_degree` solves that problem, but it found it very difficult to make progress after that as there's another copy of `P.nat_degree` hiding in the `dim` of the basis.